iceoryx_hoofs/include/iceoryx_hoofs/log/logstream.hpp

```diff
@@ -14,158 +14,115 @@
-
-    template <typename T, typename std::enable_if<std::is_arithmetic<T>::value, int>::type = 0>
-    LogStream& operator<<(const T val) noexcept
-    {
-        m_logEntry.message.append(cxx::convert::toString(val));
-        m_flushed = false;
-        return *this;
-    }
-
-    template <typename T, typename std::enable_if<std::is_base_of<LogHex, T>::value, int>::type = 0>
-    LogStream& operator<<(const T val) noexcept
-    {
-        std::stringstream ss;
-        // the '+val' is there to not interpret the uint8_t as char and print the character instead of the hex value
-        ss << "0x" << std::hex << +val.value;
-        m_logEntry.message.append(ss.str());
-        m_flushed = false;
-        return *this;
-    }
-
-    template <typename T, typename std::enable_if<std::is_base_of<LogBin, T>::value, int>::type = 0>
-    LogStream& operator<<(const T val) noexcept
-    {
-        m_logEntry.message.append("0b");
-        m_logEntry.message.append(std::bitset<std::numeric_limits<decltype(val.value)>::digits>(val.value).to_string());
-        m_flushed = false;
-        return *this;
-    }
-
-    LogStream& operator<<(const LogRawBuffer& value) noexcept;
+    inline LogStream(const char* file, const int line, const char* function, LogLevel logLevel) noexcept;
+
+    inline virtual ~LogStream() noexcept;
+
+    LogStream(const LogStream&) = delete;
+    LogStream(LogStream&&) = delete;
+
+    LogStream& operator=(const LogStream&) = delete;
+    LogStream& operator=(LogStream&&) = delete;
+
+    inline void flush() noexcept;
+
+    inline LogStream& self() noexcept;
+
+    inline LogStream& operator<<(const char* cstr) noexcept;
+
+    /// @todo iox-#1345 instead of using std::string we could also accept everything with a c_str() method
+    /// and avoid the std::string dependency
+    inline LogStream& operator<<(const std::string& str) noexcept;
+
+    template <typename T, typename std::enable_if_t<std::is_signed<T>::value, int> = 0>
+    inline LogStream& operator<<(const T val) noexcept;
+
+    template <typename T, typename std::enable_if_t<std::is_unsigned<T>::value, int> = 0>
+    inline LogStream& operator<<(const T val) noexcept;
+
+    template <typename T, typename std::enable_if_t<std::is_signed<T>::value, int> = 0>
+    inline LogStream& operator<<(const LogHex<T>&& val) noexcept;
+
+    template <typename T, typename std::enable_if_t<std::is_unsigned<T>::value, int> = 0>
+    inline LogStream& operator<<(const LogHex<T>&& val) noexcept;
+
+    template <typename T, typename std::enable_if_t<std::is_signed<T>::value, int> = 0>
+    inline LogStream& operator<<(const LogOct<T>&& val) noexcept;
+
+    template <typename T, typename std::enable_if_t<std::is_unsigned<T>::value, int> = 0>
+    inline LogStream& operator<<(const LogOct<T>&& val) noexcept;
+
+    /// @code
+    /// IOX_LOG(INFO) << "#### Hello " << [] (auto& stream) -> auto& { stream << "World"; return stream; };
```

Comment:
@MatthiasKillat this is for example useful when you want to loop through an array and want to output everything with the same log message. This would result in multiple lines

```cpp
for(const auto& num: {13, 37, 42, 73}) {
    IOX_LOG(INFO) << num << " ";
}
```

While this results in one line

```cpp
OX_LOG(INFO) << [] (auto& stream) -> auto& {
    for(const auto& num: {13, 37, 42, 73}) {
        stream << num << " ";
    }
};
```

I need to improve the documentation but this will happen with a follow up PR before this one is merged

Reply:
Yes, can be useful as syntactic sugar. The former is easier on the eyes and easier to grasp I think but the latter is more verbose but may be used to generate only one log entry.

Personally, while I like functional programming principles, I think they are often cumbersome to use in languages that are not designed for it.